pkg/streamingpromql/operators/binary_operation.go

```diff
@@ -665,22 +919,99 @@ func (b *binaryOperationSeriesBuffer) close() {
 	}
 }
 
-type binaryOperationFunc func(left, right float64) float64
+type binaryOperationFunc func(lhs, rhs float64, hlhs, hrhs *histogram.FloatHistogram) (float64, *histogram.FloatHistogram, bool, error)
 
+// FIXME(jhesketh): Investigate avoiding copying histograms for binary ops.
 var arithmeticOperationFuncs = map[parser.ItemType]binaryOperationFunc{
-	parser.ADD: func(left, right float64) float64 {
-		return left + right
+	parser.ADD: func(lhs, rhs float64, hlhs, hrhs *histogram.FloatHistogram) (float64, *histogram.FloatHistogram, bool, error) {
+		if hlhs != nil && hrhs != nil {
+			res, err := hlhs.Copy().Add(hrhs)
+			if err != nil {
+				return 0, nil, false, err
+			}
+			return 0, res.Compact(0), true, nil
+		}
+		return lhs + rhs, nil, true, nil
+	},
+	parser.SUB: func(lhs, rhs float64, hlhs, hrhs *histogram.FloatHistogram) (float64, *histogram.FloatHistogram, bool, error) {
+		if hlhs != nil && hrhs != nil {
+			res, err := hlhs.Copy().Sub(hrhs)
+			if err != nil {
+				return 0, nil, false, err
+			}
+			return 0, res.Compact(0), true, nil
+		}
+		return lhs - rhs, nil, true, nil
+	},
+	parser.MUL: func(lhs, rhs float64, hlhs, hrhs *histogram.FloatHistogram) (float64, *histogram.FloatHistogram, bool, error) {
+		if hlhs != nil && hrhs == nil {
+			return 0, hlhs.Copy().Mul(rhs), true, nil
+		}
+		if hlhs == nil && hrhs != nil {
+			return 0, hrhs.Copy().Mul(lhs), true, nil
+		}
+		return lhs * rhs, nil, true, nil
+	},
+	parser.DIV: func(lhs, rhs float64, hlhs, hrhs *histogram.FloatHistogram) (float64, *histogram.FloatHistogram, bool, error) {
+		if hlhs != nil && hrhs == nil {
+			return 0, hlhs.Copy().Div(rhs), true, nil
+		}
+		return lhs / rhs, nil, true, nil
 	},
-	parser.SUB: func(left, right float64) float64 {
-		return left - right
+	parser.POW: func(lhs, rhs float64, hlhs, hrhs *histogram.FloatHistogram) (float64, *histogram.FloatHistogram, bool, error) {
+		return math.Pow(lhs, rhs), nil, true, nil
 	},
-	parser.MUL: func(left, right float64) float64 {
-		return left * right
+	parser.MOD: func(lhs, rhs float64, hlhs, hrhs *histogram.FloatHistogram) (float64, *histogram.FloatHistogram, bool, error) {
+		return math.Mod(lhs, rhs), nil, true, nil
 	},
-	parser.DIV: func(left, right float64) float64 {
-		return left / right
+	parser.ATAN2: func(lhs, rhs float64, hlhs, hrhs *histogram.FloatHistogram) (float64, *histogram.FloatHistogram, bool, error) {
+		return math.Atan2(lhs, rhs), nil, true, nil
 	},
-	parser.MOD:   math.Mod,
-	parser.POW:   math.Pow,
-	parser.ATAN2: math.Atan2,
+}
+
+// dataSeries couples data, sourceSeriesIndices, and sourceSeriesMetadata for use by mergeOneSide
+// Specifically it implements sort.Interface which will sort both data and sourceSeriesIndices at the same
+// time to retain a mapping of their indexes on either the present float timestamps or histograms as determined by sortByFloat.
+type dataSeries struct {
```

Comment:
I also think we should split this in three, to separate the data and the sorting of it: one thing that holds the data, and then two implementations of `sort.Interface` that hold a reference to the data, one for sorting based on float samples and the other for sorting based on histogram series.

Reply:
Say we have `seriesForOneGroupSideFloats` and `seriesForOneGroupSideHistograms`, both with `data` and `sourceSeriesIndices`.

The underlying slices for `data` and `sourceSeriesIndices` are the same. So when you sort `seriesForOneGroupSideFloats` it'll actually re-sort `seriesForOneGroupSideHistograms` as well.

By coupling with a toggle like this I feel it makes the relationship tighter and clearer.

Reply:
I should have been clearer sorry: I'm suggesting having one thing (say `seriesForOneGroupSide`) that holds the data, and then two other structs that are just the `sort.Interface` implementations that take a reference to the `seriesForOneGroupSide`, for example:

```golang
type seriesForOneGroupSide struct {
	data                 []types.InstantVectorSeriesData
	sourceSeriesIndices  []int
}

type floatSideSorter struct {
	side seriesForOneGroupSide
}

type histogramSideSorter struct {
	side seriesForOneGroupSide
}
```

Reply:
ah, I see, okay, that's reasonable